src/vellum/workflows/expressions/tests/test_contains.py

```
@@ -0,0 +1,178 @@
+import pytest
+
+from vellum.workflows.constants import undefined
+from vellum.workflows.descriptors.exceptions import InvalidExpressionException
+from vellum.workflows.expressions.contains import ContainsExpression
+from vellum.workflows.references.constant import ConstantValueReference
+from vellum.workflows.state.base import BaseState
+
+
+class TestState(BaseState):
+    dict_value: dict = {"key": "value"}
+    list_value: list = [1, 2, 3]
+    string_value: str = "hello world"
+
+
+def test_dict_contains_identical_dict():
+    """
+    Tests that ContainsExpression handles dict-contains-dict scenarios.
+
+    This addresses Linear ticket APO-1025 where dict.contains(dict)
+    was failing with 'unhashable type' error.
+    """
+    state = TestState()
+    lhs_dict = {"foo": "bar"}
+    rhs_dict = {"foo": "bar"}
+
+    expression = ContainsExpression(lhs=lhs_dict, rhs=rhs_dict)
+    result = expression.resolve(state)
+
+    assert result is True
+
+
+def test_dict_contains_different_dict():
+    """
+    Tests that ContainsExpression returns False for different dicts.
+    """
+    state = TestState()
+    lhs_dict = {"foo": "bar"}
+    rhs_dict = {"hello": "world"}
+
+    expression = ContainsExpression(lhs=lhs_dict, rhs=rhs_dict)
+    result = expression.resolve(state)
+
+    assert result is False
+
+
+def test_string_contains_dict_as_json():
+    """
+    Tests that ContainsExpression handles string-contains-dict scenarios.
+    """
+    state = TestState()
+    lhs_string = 'Response: {"status": "success"} was returned'
+    rhs_dict = {"status": "success"}
+
+    expression = ContainsExpression(lhs=lhs_string, rhs=rhs_dict)
+    result = expression.resolve(state)
+
+    assert result is True
```

[blocking] this is a bit unintuitive both in python and typescript this is either false or error.

We should be fine continuing to raise an error here, as long as it makes it more clear to the end user what's happening instead of the current "Internal Server Error".

If `rhs` is a dict (or in general, any incompatible type for the left side of the ` in ` expression), I think we should raise an `InvalidExpressionException`. Our node logic properly catches and propagates this error so that the end user could see their error more easily and remedy accordingly

---

added